### PR TITLE
Added a pre-build-script option to MooseDocs build

### LINKED ISSUE #23900: Allow MooseDocs to execute plot scripts
## Reason
This will allow plots to be generated that can be used in documentation, for instance, plots of assessment cases. This will support #21133, for example, as well as many applications.

## Design
See #23912 for the evolving design.

## Impact
New option.

### PR DESCRIPTION
**Pull request by joshuahansel**
This is a draft of the basic idea so far. Testing and doc to come if this design is ok.

Closes #23900

### CODE REVIEW

**Review by loganharbour — COMMENTED** [review]
This needs to be a config option because it must be ran ahead of time every time someone wants to build, otherwise we'll get media include failures

**Comment by cticenhour** [comment]
If infrastructure is added to run this as part of the media extension (for example), then it could run while the website is being built, giving output to the screen about what it is doing, and then making it clear to a reader of the markdown what media is generated using a script. Perhaps something like:

```
!media script my_fig.py
```

No build option would then be required. At minimum it would take in one script for one figure, but Logan and I could both see a scenario where you could list multiple scripts to make a gallery or something like that!

**Comment by joshuahansel** [comment]
Ok, you mean replacing
```
!media my_plot.png
```
with
```
!media script my_plot_script.py
```

This could work. I know a lot of my plot scripts right now do multiple plots in the same script, but I could probably live with breaking them up.

Under the hood, it's still working with an image, right? How might we get the name of that image?

**Comment by cticenhour** [comment]
>```
> !media script my_plot_script.py
> ```
> This could work. I know a lot of my plot scripts right now do multiple plots in the same script, but I could probably live 
> with breaking them up.

I don't see a problem with having the `script` subcommand be a trigger for multiple image files in principle, if we come up with a good way to preserve generality. If a script generates multiple plots, then multiple tokenizing and rendering steps could be taken for the generated images so that they are seen on the page in the location you have placed the call to the extension. This, of course, would be an issue if your script generates multiple images that belong to different locations on the page. At the end of the day, I want to preserve the readability/flow of the markdown page. When you call the media extension to insert a video, the video ends up in that spot on the page relative to what is around it. If the `media script` subcommand is called, then everything generated by that call should go in the location where it is called. 

> Maybe we could have the plot script return the name of the plot? Otherwise, it wouldn't be the end of the world to do
> ```
> !media script my_plot_script.py my_plot.png
> ```

Thinking about this a bit, I personally like the idea of this syntax meaning that the system should *expect* `my_plot.png` and it fails when that isn't seen. Just to be clear, are you thinking this should mean the file that leaves `my_plot_script.py` is renamed to `my_plot.png`? The latter almost implies to me that we need to be able to pass things in the script, when I think we should probably stay away from that. The script should just be able to run IMO. I am in favor of the former idea....if, for whatever reason, the script completes and `my_plot.png` doesn't exist then the user should get a good error message. 

### Thinking out loud a bit after this point about the whole idea of scripts in docs...

Inserting scripts into the docs system does add complexity with respect to python package requirements (or otherwise, if we want to support non-python scripts). Python will produce an error whenever a package is missing, which is awesome for our CI. So then we add a new package to `moose-tools` to support that new image script....and on and on as scripts are added. That could end up with a very bloated conda package that changes quite a lot. What do you think?

**Comment by joshuahansel** [comment]
I'm on board with the `!media script` idea because I see that advantage of running a script when you need it - I guess this allows it to be more parallel? And then you don't risk generating any plots that you don't actually use anywhere.

As far as multiple plots in a script, I'd vote to just keep it simple: one plot per script.

I think `!media script myscript.py myplot.png` is good.

I think the possibility of missing dependencies isn't too much of an issue. I assume the conda package has `matplotlib` and `numpy`? Those are the only two I can think of. I also rely on some functions in `thermal_hydraulics/python`, but we can always add to PYTHONPATH. If we get an error generating documentation, then that's fine - the developer can request dependencies or adjust to not need them.

**Comment by cticenhour** [comment]
I am in agreement with all of this. My only outstanding design question is what you want the presence of a `.png` file name to mean in `!media script myscript.py myplot.png'? Is it an output rename for consistency, or is it a way to error check that the script produced the file we want it to?

**Comment by joshuahansel** [comment]
> I am in agreement with all of this. My only outstanding design question is what you want the presence of a `.png` file name to mean in `!media script myscript.py myplot.png`? Is it an output rename for consistency, or is it a way to error check that the script produced the file we want it to?

Not an output rename - just the name of the plot that should be generated. Check for it and error if it doesn't exist after running the script.

**Comment by cticenhour** [comment]
Sounds great to me!

**Comment by joshuahansel** [comment]
@loganharbour @cticenhour This iteration works... but only if I use the serial executioner:
```
cd python/MooseDocs/test
MOOSE_DIR=../../.. ./moosedocs.py build --serve --executioner MooseDocs.base.Serial
```

I'm not exactly sure what needs to be done to fix it yet, but I just wanted to get opinions on the design here. I did this
```
!media plot.jpg plot_script=make_plot.py
```
instead of the original idea
```
!media script make_plot.py plot.jpg
```
because it made more sense code-wise. Thoughts?

Also, I made the assumption that the generated plot resides in the same directory as the plot script because I wasn't sure how to do it otherwise.

**Comment by cticenhour** [comment]
Based on the structure of the media extension, I think your choice for syntax makes sense here, and I am OK with it.

Re: the executioner only allowing serial mode.....I wonder if its the call to the `subprocess` package to run the script that is mucking things up? Running these scripts as a "pre-render" step (since I *think* those are run serially when rendering is done in parallel) might be an answer for this, though we might have to re-think how we set things up. Pulling the guts of your changes out to a new pre-render method might mean that having it in the media extension no longer makes sense, and that having a list of scripts (provided via `config.yml` maybe) that need to be run and output files added to the content before renders occur might be "better". Thinking out loud here if you can't get the current set of changes to work with minor changes.

**Comment by joshuahansel** [comment]
> Based on the structure of the media extension, I think your choice for syntax makes sense here, and I am OK with it.
> 
> Re: the executioner only allowing serial mode.....I wonder if its the call to the `subprocess` package to run the script that is mucking things up? Running these scripts as a "pre-render" step (since I _think_ those are run serially when rendering is done in parallel) might be an answer for this, though we might have to re-think how we set things up. Pulling the guts of your changes out to a new pre-render method might mean that having it in the media extension no longer makes sense, and that having a list of scripts (provided via `config.yml` maybe) that need to be run and output files added to the content before renders occur might be "better". Thinking out loud here if you can't get the current set of changes to work with minor changes.

To clarify, what's happening in parallel is that the new (plot) page is not being found. I don't know how the parallel execution works here. I guess maybe I added the page on one (thread? processor? I don't know) and then it wasn't found on another. 🤷‍♂️

**Comment by joshuahansel** [comment]
Is there anyone that is pretty comfortable with development within MooseDocs other than Andrew? 😜

**Comment by cticenhour** [comment]
I can make an attempt at sorting it out, since I need to get better at this stuff anyway. If this can wait, I can take a look at it after I get back from North Carolina (so in about two weeks). If you want this off the PR list more quickly, Andrew hasn't minded chiming in with suggestions  in the past if you want to tag him.

File: python/MooseDocs/extensions/media.py [hunk 1]
```diff
@@ -79,13 +80,43 @@ class ImageCommand(command.CommandComponent):
...
+            if not os.path.isfile(plot_path):
+                LOG.error("The plot script '%s' must generate the plot '%s'", script_localname, plot_localname)
+
+            # Add the plot page if the plot did not exist before building
+            print(plot_localname)
+            if not self.translator.findPage(plot_localname, throw_on_zero=False, exact=False):
+                plot_page = pages.File(plot_localname, source=plot_path)
+                self.translator.addPage(plot_page)
+                self.translator.initPage(plot_page)
+                print("Added")
```
Comment (joshuahansel) [inline python/MooseDocs/extensions/media.py:118]: Just a reminder to myself to remove these print statements...

**Comment by joshuahansel** [comment]
It's not super urgent, but it'd be nice to get it done soon.

@aeslaughter Sorry to haunt you with a MooseDocs question... blame Casey! If this takes more than a few minutes of your time, please feel free to decline :D

I'm basically trying to modify the media extension to allow a plot to be generated on the fly from a script. It works for the serial executioner but not the default parallel one (whichever one it is). You can see the code, but basically, I have the `!media` command call the script and then immediately add the plot "page". But in parallel execution, this page isn't found later. Am I able to fix this?

**Comment by joshuahansel** [comment]
Not urgent, but I am still interested in this capability. @cticenhour if you ever find yourself diving deep into MooseDocs, could you think about this? Otherwise, don't prioritize it. If we can't figure it out, I might advocate for the original design again...

**Comment by cticenhour** [comment]
I can put it on my list, as I've got other moosedocs things to take care of. Can't promise anything until next FY though...

**Comment by cticenhour** [comment]
The parallelism work I'm doing for another moosedocs project for the VTB may come in handy with the remaining issues we have on this version of the code.

**Comment by aeslaughter** [comment]
If you want you can setup a Teams call with me and I can see if I can help you figure this out. If I was trying to figure out what was going wrong I would add this new script thing to the test document and work there, it is much faster to build/debug.

**Comment by joshuahansel** [comment]
@aeslaughter Thanks! I knew you missed moosedocs. I think I had been using the test doc as my test ground - definitely faster! And @cticenhour just let me know when you're ready for that, and I can join.

**Comment by aeslaughter** [comment]
@joshuahansel I just miss you

**Comment by joshuahansel** [comment]
Closing this PR in favor of https://github.com/idaholab/moose/pull/25241

**Pull request closed without merging**